Detect mismatch of execution spaces in functors

masterleinad opened the pull request:
Fixes #2074.

masterleinad committed:
Detect mismatch of execution spaces in functors

ldh4 reviewed (approved):
I was in a process of converting my initial solution, similar to this, into a runtime check in attempt to make an unit-test, but I think compile-time check is sufficient, given that it is a fairly simple check.

dalg24 commented on core/src/Kokkos_Parallel.hpp:
```diff
@@ -87,6 +87,31 @@ using device_type_t = typename T::device_type;
+          std::is_same<functor_execution_space, nonesuch>::value ||
+          std::is_same<policy_execution_space, functor_execution_space>::value,
+      "A policy with an execution space and a functor with an execution space "
+      "are given but the execution space types don't match!");
```
```suggestion
      "are given but the execution space types do not match!");
```

dalg24 reviewed (commented):
I would prefer if you used `is_detected<execution_space_t, Policy>::value` instead of `std::is_same<detected_t<execution_space_t, Policy>, nonesuch>::value`

masterleinad committed:
don't -> do not

masterleinad committed:
Avoid nonesuch

masterleinad commented:
@dalg24 I addressed your suggestions. 

In my opinion, the changes here would be a good example of why defining `detected_or` would be convenient. I could use the result both for checking that the check was successful and get the referenced type in case of success.

dalg24 commented:
> In my opinion, the changes here would be a good example of why defining detected_or would be convenient. I could use the result both for checking that the check was successful and get the referenced type in case of success.

Fair enough.  I still don't like the idea of writing
```C++
using FunctorExecutionSpace = detected_or<MyDefault, execution_space_t, Functor>;
static_assert(!typename FunctorExecutionSpace::value_type::value || ...);
// or
static_assert(!std::is_same<MyDefault, typename FunctorExecutionSpace::type>::value || ...);
